### PR TITLE
Numba lowmem

### PR DESCRIPTION
**Pull request by meiertgrootes**
This compiles correctly and passes tests, but requesting review and discussion before merge

### CODE REVIEW

File: cgc/coclustering.py [hunk 1]
```diff
@@ -33,8 +33,17 @@ class Coclustering(object):
...
+    def __init__(self,
+                 Z,
+                 nclusters_row,
+                 nclusters_col,
+                 conv_threshold=1.e-5,
+                 max_iterations=1,
+                 nruns=1,
+                 epsilon=1.e-8,
+                 low_memory=False,
+                 numba_jit=False,
```
Comment (rogerkuou) [inline cgc/coclustering.py:45]:
```suggestion
```
@meiertgrootes Hi Meiert the `numba_jit` is the argument for `run_threads` right? Maybe we do not need to add it here.

File: cgc/coclustering.py [hunk 2]
```diff
@@ -33,8 +33,17 @@ class Coclustering(object):
...
-                 max_iterations=1, nruns=1, epsilon=1.e-8, output_filename=''):
+    def __init__(self,
+                 Z,
+                 nclusters_row,
+                 nclusters_col,
+                 conv_threshold=1.e-5,
+                 max_iterations=1,
+                 nruns=1,
+                 epsilon=1.e-8,
+                 low_memory=False,
```
Comment (rogerkuou) [inline cgc/coclustering.py:44]:
```suggestion
```

File: cgc/coclustering.py [hunk 3]
```diff
@@ -167,23 +186,25 @@ def _dask_runs_performance(self, row_clusters=None, col_clusters=None):
...
+        row_min, col_min, e_min = None, None, 0.
+        r = 0
         for future, result in dask.distributed.as_completed(
-                futures,
-                with_results=True,
-                raise_errors=False):
-            logger.info(f'Retrieving run {self.results.nruns_completed} ..')
+                futures, with_results=True, raise_errors=False):
+            logger.info(f'Waiting for run {r} ..')
+
```
Comment (rogerkuou) [inline cgc/coclustering.py:207]:
```suggestion
                futures,	                
                with_results=True,	            
                raise_errors=False):	
                logger.info(f'Retrieving run {self.results.nruns_completed} ..')
```

Comment (rogerkuou) [reply cgc/coclustering.py:207]: Here I kept the implementation of the dev branch

File: cgc/coclustering_numpy.py [hunk 4]
```diff
@@ -12,8 +13,46 @@ def _distance(Z, X, Y, epsilon):
...
+        part1[i, :] = sum_part1
+
+    part2 = Z.shape[1] * epsilon
+    part3 = np.dot(Z, np.log(product + epsilon))
+    return part1 + part2 - part3
+
+
 def _initialize_clusters(n_el, n_clusters):
-    """ Initialize cluster array """
+    """ Initialize cluster occupation matrix """
```
Comment (rogerkuou) [inline cgc/coclustering_numpy.py:55]:
```suggestion
    """ Initialize cluster array """
```

File: cgc/coclustering_numpy.py [hunk 5]
```diff
@@ -57,20 +150,54 @@ def coclustering(Z, nclusters_row, nclusters_col, errobj, niters, epsilon,
...
-        CoCavg = (np.dot(np.dot(R.T, Z), C) + Gavg * epsilon) / (
-                np.dot(np.dot(R.T, np.ones((m, n))), C) + epsilon)
+        if low_memory:
+            if numba_jit:
+                CoCavg = (_cluster_dot_numba(Z, row_clusters, col_clusters,
+                                             nclusters_row, nclusters_col) +
+                          Gavg * epsilon) / (_cluster_dot_numba(
+                              np.ones((m, n)), row_clusters, col_clusters,
+                              nclusters_row, nclusters_col) + epsilon)
+
```
Comment (rogerkuou) [inline cgc/coclustering_numpy.py:160]:
```suggestion
```

File: cgc/coclustering_numpy.py [hunk 6]
```diff
@@ -57,20 +150,54 @@ def coclustering(Z, nclusters_row, nclusters_col, errobj, niters, epsilon,
...
+                d_row = _distance_lowmem_numba(Z, col_clusters, CoCavg.T,
+                                               epsilon)
+            else:
+                d_row = _distance_lowmem(Z, col_clusters, CoCavg.T, epsilon)
+        else:
+            d_row = _distance(Z, np.ones((m, n)), np.dot(C, CoCavg.T), epsilon)
         # Assign to best row cluster
         row_clusters = np.argmin(d_row, axis=1)
-        R = _setup_cluster_matrix(nclusters_row, row_clusters)
+
```
Comment (rogerkuou) [inline cgc/coclustering_numpy.py:184]:
```suggestion
```

File: geoclustering/coclustering_numpy_lowmem.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
+import numpy as np
```
Comment (rogerkuou) [inline geoclustering/coclustering_numpy_lowmem.py:1]: We can remove this file?

Comment (meiertgrootes) [reply geoclustering/coclustering_numpy_lowmem.py:1]: yes, this can go. It's legacy from before number

File: cgc/coclustering.py (hunk 1 above)
Comment (meiertgrootes) [reply cgc/coclustering.py:45]: technically its no to run threads, but within coclustering (in the numpy version).

**Review by rogerkuou — CHANGES_REQUESTED** [review]
Hi @meiertgrootes and @fnattino, I updated the pull request. Now the conflicts are solved and all tests passed. 

The differences of low_mem and normal numpy run has been solved. Indeed it was caused by the way of initializing the clusters. I added a unit test (`test_run_with_threads_lowmem`) in `test_coclustering` to guarantee this.

I also commented some changes which I am not fully sure. Could you please give a last check?

File: cgc/coclustering.py (hunk 1 above)
Comment (meiertgrootes) [reply cgc/coclustering.py:45]: Sorry, I see what you meant now. Yes, that is fine.

**Review by rogerkuou — APPROVED** [review]

**Comment by rogerkuou** [comment]
I adapted the unit test regarding the coclustering numpy. I think it is all good now. will merge this

**Pull request merged**